ocaml/testsuite/tests/pattern-guards/test.ml

```diff
@@ -260,3 +260,38 @@ say_hello_catching_exns None name_map;;
 [%%expect{|
 - : string = "Hello, Fred"
 |}]
+
+let patch_to_match_failure f default x =
+  match x with
+  | None -> default
+  | Some x when f x match Some y -> y
+;;
+
+let exact_half n = if n mod 2 = 0 then Some (n / 2) else None;;
+[%%expect{|
+Lines 2-4, characters 2-37:
+2 | ..match x with
+3 |   | None -> default
+4 |   | Some x when f x match Some y -> y
+Warning 8 [partial-match]: this pattern-matching is not exhaustive.
+Here is an example of a case that is not matched:
+Some _
+(However, some guarded clause may match this value.)
+val patch_to_match_failure : ('a -> 'b option) -> 'b -> 'a option -> 'b =
+  <fun>
+val exact_half : int -> int option = <fun>
+|}];;
+
+patch_to_match_failure exact_half ~-1 None;;
+[%%expect{|
+- : int = -1
+|}];;
+patch_to_match_failure exact_half ~-1 (Some 42);;
+[%%expect{|
+- : int = 21
+|}];;
+patch_to_match_failure exact_half ~-1 (Some 41);;
+[%%expect{|
+Exception: Match_failure ("", 2, 2).
+|}];;
+
```

Comment:
I realize we don't have a test for warning 57 for pattern guards. Would you mind adding one? I don't think it's necessarily related to this PR, but the fact that it involves or-patterns and guards sets off alarm bells in my mind.

Reply:
Done! Also added one for warning 8 (which we weren't testing explicitly / in the most interesting case)

Reply:
Could you clean up the output a bit?
* Insert more `%%expect` blocks so the output is right under the definition.
* As useful, add comments explaining why you've added certain tests

Reply:
Did both of these things, let me know if there are ways to further improve test output quality.